crates/terminal/src/terminal.rs

```diff
@@ -1425,6 +1417,92 @@ impl Terminal {
         }
         Task::ready(())
     }
+
+    fn register_task_finished(
+        &mut self,
+        error_code: Option<i32>,
+        cx: &mut ModelContext<'_, Terminal>,
+    ) {
+        self.completion_tx.try_send(()).ok();
+        let task = match &mut self.task {
+            Some(task) => task,
+            None => {
+                if error_code.is_none() {
+                    cx.emit(Event::CloseTerminal);
+                }
+                return;
+            }
+        };
+        if task.status != TaskStatus::Running {
+            return;
+        }
+        match error_code {
+            Some(error_code) => {
+                task.status.register_task_exit(error_code);
+            }
+            None => {
+                task.status.register_terminal_exit();
+            }
+        };
+        let escaped_full_label = task.full_label.replace("\r\n", "\r").replace('\n', "\r");
```

Comment:
Could we hoist out label processing into a helper function?

Reply:
Sure, it's a bit oddly typed (`-> (String, String)`) as a small desire to avoid extra allocations, but I can rewrite it to be more readable, if needed.